base/compiler/ssair/inlining.jl

```diff
@@ -1068,8 +1081,11 @@ function narrow_opaque_closure!(ir::IRCode, stmt::Expr, @nospecialize(info), sta
             # N.B.: Narrowing the ub requires a backdge on the mi whose type
             # information we're using, since a change in that function may
             # invalidate ub result.
-            push!(state.et, unspec_call_info.mi)
-            stmt.args[4] = newT
+            mi = Core.Compiler.specialize_method(unspec_call_info.match, true)
+            if mi !== nothing
```

Comment:
Why force `preexisting`? It is a bit odd appearing already that it isn't using the `mi` object, since someday these edges may get refined to describe why they exist, and this one doesn't seem to provide any information

(actually, I'm just skeptical there is a new backedge here, and not already the one from inference suffices)

Reply:
Well, during inference there isn't really a backedge, because we never actually properly call the thing. Of course at the moment, we're adding the backedge anyway, but I think conceptually, this is where the backedge is really being introduced. Basically what needs to happen here is that this optimized result gets invalidated if the return type of the opaque closure method itself could change, because that would make the narrowing no longer legal. I think that implies that this needs to be pre-existing, since otherwise, why would we have a meaningful return type here?

Reply:
That's basically right, but reversed: There is an edge during inference, since those edges are to the types, not to a call. For example, `return_type` has proper edges added there, so should have nothing additional to do here. This new object is the same. Then here, in optimization, we introduce edges to the calls themselves. The former (inference type-edges) tends to be a strict superset of the latter (optimization call-edges), while the later are a strict refinement of the former (the code is more strict than the type it computed). There is no call here, so during optimization there isn't really a new backedge, since there is no call (just the existing edge to the type computation). See?

Reply:
> There is an edge during inference, since those edges are to the types

Sure, but we're not actually using the type during inference, right? So there doesn't really need to be an edge here during inference. The only reason we do that inference call at all is to be able to use it here - otherwise both the inference result and the optimization are entirely unaffected.

Reply:
Vaguely not, but I think the same could be said of any unused call result too. I think the other place that can consume this type is `tmerge`—though we sometimes instead reject tracking edges that could make a result more specific. We attempt a small amount of DCE to try to trim a few of those edges, but currently we just mostly don't track the effect of the edge (which someday we perhaps should do).